File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount == 0) {
            return 0;
        }

        vector<int> dp(amount + 1, INT_MAX);

        for (int coin : coins) {
            if (coin <= amount) {
                 dp[coin] = 1;
            }
           
        } 

        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int j = 0; j < coins.size(); j++) {
                if (coins[j] <= amount && i - coins[j] >= 0 && dp[i - coins[j]] != INT_MAX) {
                    dp[i] = min(dp[i], dp[i - coins[j]] + 1); 

                }
            }
        }

        return dp[amount] == INT_MAX ? -1 : dp[amount];
    }
};

File: 0322-coin-change/0322-coin-change.ts
function coinChange(coins: number[], amount: number): number {
    let dp : number[] = Array(amount + 1).fill(Number.MAX_SAFE_INTEGER);
    dp[0] = 0;
    
    for (let i = 1; i <= amount; i++) {
        for (let j = 0; j < coins.length; j++) {
            if (i >= coins[j]) {
                dp[i] = Math.min(dp[i], dp[i - coins[j]] + 1);
            }
        }
    }
    
    return dp[amount] === Number.MAX_SAFE_INTEGER ? -1 : dp[amount];
};